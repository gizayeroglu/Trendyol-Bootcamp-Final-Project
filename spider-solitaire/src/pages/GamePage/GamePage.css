:root {
  --blackColor:#000;
  --iconColor: #ffb142;
}

.card-holder-container {
  display: flex;
  flex-direction: column;
}
.game-area-container {
  display: grid;
  grid-template-columns: repeat(10, 1fr);
  justify-content: center;
  justify-items: center;
  margin-top: 150px; /*************************************************************************************************/
}
.game-completed-card-area {
  height: 105px;
	width: 75px;
	border-radius: 10px;
	border: 2px solid var(--blackColor);
}
.game-area-top-containers {
  display: flex;
  flex-direction: row;
  gap: 20px;
  align-content: center;
  justify-content: flex-end;
  margin: 30px 30px;
  margin-top: -50px;
}
.score i,
.gameTime i { 
  color:var(--iconColor);
  font-size: 20px;
}
.score,
.gameTime { 
  font-size: 17px;
  font-weight: 600;
  font-style: italic;
}
.gameOutputs {
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 100px;
  border-bottom: 2px solid var(--iconColor);
  padding: 10px;
  margin-top: 10px;
}