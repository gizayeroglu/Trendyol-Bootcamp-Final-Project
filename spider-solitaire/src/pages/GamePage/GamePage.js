import './GamePage.css';

import React, { useState, useRef, useEffect } from 'react';
import GoBack from '../../components/GoBack/GoBack';
import CardHolder from '../../components/CardHolder/CardHolder';
import Card from '../../components/Card/Card';
import CardDistributor from '../../components/CardDistributor/CardDistributor';
import TimeUpCounter from '../../components/TimeUpCounter/TimeUpCounter';
import { getCardHoldersWithCards, checkSetOfCards, isAnyCardHolderWithoutCards, updateCardDraggable, isValidDrop, getHintedData } from '../../utils/GameUtils';

const data = getCardHoldersWithCards(54);

function GamePage() {
  const [deckData, setDeckData] = useState(data);
  const [gameScore, setGameScore] = useState(0);
  const [highestScore, setHighestScore] = useState(0);
  const [lastMove, setLastMove] = useState({});
  const [isLastCardOpenedWhenDrop, setIsLastCardOpenedWhenDrop] = useState(false);

  const dragCard = useRef();
  const dragCardHolder = useRef();


  const handleDragStart = (e, { cardHolderIndex, cardIndex }) => {
    dragCard.current = {x:e.pageX, y:e.pageY, targetCard:e.target, cardHolderIndex, cardIndex };
    e.dataTransfer.setDragImage(new Image("0", "0"), -10, -10);
    dragCardHolder.current = e.target;
    dragCardHolder.current.addEventListener('drag', handleDrag);
  };

  const handleDrag = (e) => {
    const { x, y, targetCard } = dragCard.current;
    const moveX = e.pageX - x;
    const moveY = e.pageY - y;

    let css;
    
    if (e.pageX === 0 ) {
      css = 
      `z-index:0;
      transform:translate(0px,0px);`;
    } else {
      css =
      `z-index:9;
      pointer-events: none;
      transform: scale(1.05, 1.05) rotate(0deg) translate(${moveX}px, ${moveY}px);
      position:relative;`
    }

    targetCard.style.cssText = css;
    let currentNode = targetCard.nextSibling;

    while (currentNode) {
      currentNode.style.cssText = css;
      currentNode = currentNode.nextSibling;
    }
  };
  
  const handleDrop = (e, { cardHolderIndex }) => { 
    e.preventDefault();
    const droppedCardHolderIndex = cardHolderIndex;
    const draggedCardHolderIndex = dragCard.current.cardHolderIndex;
    const draggedCardIndex = dragCard.current.cardIndex;
    const targetCard =  dragCard.current.targetCard;

    const isValid = isValidDrop(deckData, droppedCardHolderIndex, draggedCardHolderIndex, draggedCardIndex);

    if(droppedCardHolderIndex === draggedCardHolderIndex || !isValid) {
      const css = 'z-index:0;pointer-events:auto;';
      targetCard.style.cssText = css;
      let currentNode = targetCard.nextSibling;

      while (currentNode) {
        currentNode.style.cssText = css;
        currentNode = currentNode.nextSibling;
      }

      return;
    }

    const newDeckData = [...deckData];

    const removedCardsCount = newDeckData[draggedCardHolderIndex].cards.length - draggedCardIndex;
    const removedCards = newDeckData[draggedCardHolderIndex].cards.splice(draggedCardIndex, removedCardsCount);

    //open last card of dragged holder
    const lastCard = newDeckData[draggedCardHolderIndex].cards[newDeckData[draggedCardHolderIndex].cards.length - 1];
    if (lastCard && !lastCard.isOpen) {
      lastCard.isOpen = true;
      lastCard.isDraggable = true;
      setIsLastCardOpenedWhenDrop(true); //if the last card opened with drag it will be return closed on undo last move
    } else if(lastCard){
      lastCard.isDraggable = true;
      setIsLastCardOpenedWhenDrop(false);
    }
    
    const { isScore, cards } = checkSetOfCards([...newDeckData[droppedCardHolderIndex].cards, ...removedCards]);
    newDeckData[droppedCardHolderIndex].cards = cards;

    const updatedCardsData = updateCardDraggable(newDeckData);
    setDeckData(updatedCardsData);

    if(isScore) {
      setGameScore(gameScore + 300);
      setLastMove({}); // If the set is completed, clean last move state
    } else {
      setLastMove({droppedCardHolderIndex, draggedCardHolderIndex, removedCardsCount});
    }

    dragCard.current = null;
  };

  const dragOver = (e) => {
    e.preventDefault();
  };

  const distributeNewCards = (e) => {
    const isExist = isAnyCardHolderWithoutCards(deckData);

    if(isExist) {
      alert('There must be at least one card in each column before you can deal a new row cards !');
      return;
    }

    e.target.style.display = 'none';
    const distrubitedData = getCardHoldersWithCards(10);
    const newDeckData = [...deckData];
    
    for (const data of distrubitedData) {
      const foundedHolder = newDeckData.find(el => el.name === data.name);

      if(foundedHolder) {
        const { isScore, cards } = checkSetOfCards([...foundedHolder.cards, ...data.cards]);
        foundedHolder.cards = cards;

        if(isScore) setGameScore(gameScore + 300);
      } 
    }
      
    const updatedCardsData = updateCardDraggable(newDeckData);
    setDeckData(updatedCardsData);
    setLastMove({}); // If the new cards distributed, clean last move state
  };
  
  const handleHint = () => {
    const hintedData = getHintedData([...deckData]);
    if(!hintedData) return alert(`I couldn't find a move either`);
      
    setDeckData(hintedData);
    
    setTimeout(()=>{
      const newData = [...hintedData];

      for (const data of newData) {
        for (const card of data.cards) {
          if(card.isHighlighted) card.isHighlighted = false;
        }
      }

      setDeckData(newData);
    }, 1000);
  };

  const handleTakeBackCardMove = () => {
    const droppedIndexBefore = lastMove.droppedCardHolderIndex;
    const draggedIndexBefore = lastMove.draggedCardHolderIndex;
    const removedCardsCount = lastMove.removedCardsCount;

    try {     
      const newDeckData = [...deckData];
      
      const droppedCards = newDeckData[droppedIndexBefore].cards;
      const restoredCards = droppedCards.splice(droppedCards.length-removedCardsCount, removedCardsCount);

      if(newDeckData[draggedIndexBefore].cards.length && isLastCardOpenedWhenDrop) {
          //close the prev card when you undo your last movement if it is opened on drag 
        newDeckData[draggedIndexBefore].cards[newDeckData[draggedIndexBefore].cards.length-1].isOpen = false;
        newDeckData[draggedIndexBefore].cards[newDeckData[draggedIndexBefore].cards.length-1].isDraggable = false;
      }

      newDeckData[draggedIndexBefore].cards = [...newDeckData[draggedIndexBefore].cards, ...restoredCards];
      
      const updatedCardsData = updateCardDraggable(newDeckData);
      setDeckData(updatedCardsData);
      setLastMove({});
    } catch (ex) {
      alert('There is no last move');
    }
  };

  useEffect(() => {
    if(gameScore === 2400){
      alert('Congrats ! ðŸ”¥');
    }
    const highestScore = localStorage.getItem('highestScore');
    if(!highestScore) localStorage.setItem('highestScore', highestScore);
    setHighestScore(localStorage.getItem('highestScore'));

    if(gameScore >= highestScore){
      localStorage.setItem('highestScore', gameScore);
      setHighestScore(gameScore);
    }
  },[gameScore]); 

  return (
    <>
      <div className='game-outputs'>
        <GoBack />
        <span className='score'>
          <i className='fas fa-trophy'></i> Score: {gameScore}
        </span>
        <span className='highest-score'>
          <i className='fas fa-dragon'></i> Highest Score: {highestScore}
        </span>
        <span className='game-time'>
          <i className='fas fa-hourglass-half'></i> <TimeUpCounter />
        </span>
        <button className='undo-move-button' onClick={handleTakeBackCardMove}>Undo Last Move</button>
        <button className='hint-button' onClick={handleHint}>Hint</button>
        <button className='refresh-page-button' onClick={() => window.location.reload(false)}>Restart</button>
      </div>
      <div className='game-area-top-containers'>
         {gameScore >= 300 ? <Card isOpen='true' symbol='K' /> : <div className='empty-card-area'></div>}
         {gameScore >= 600 ? <Card isOpen='true' symbol='K' /> : <div className='empty-card-area'></div>}
         {gameScore >= 900 ? <Card isOpen='true' symbol='K' /> : <div className='empty-card-area'></div>}
         {gameScore >= 1200 ? <Card isOpen='true' symbol='K' /> : <div className='empty-card-area'></div>}
         {gameScore >= 1500 ? <Card isOpen='true' symbol='K' /> : <div className='empty-card-area'></div>}
         {gameScore >= 1800 ? <Card isOpen='true' symbol='K' /> : <div className='empty-card-area'></div>}
         {gameScore >= 2100 ? <Card isOpen='true' symbol='K' /> : <div className='empty-card-area'></div>}
         {gameScore >= 2400 ? <Card isOpen='true' symbol='K' /> : <div className='empty-card-area'></div>}
        <CardDistributor  getNewCards={distributeNewCards} />
      </div>
      <div className='game-area-container'>
        {deckData.map((cardHolder, cardHolderIndex) => {
          return (
            <CardHolder key={cardHolder.name} onDrop={(e) => handleDrop(e, { cardHolderIndex })} onDragOver={dragOver}>
              {cardHolder.cards.map((card, cardIndex) => {
                return (
                  <Card
                    key={`${cardHolderIndex}-${cardIndex}`}
                    onDragStart={(e) => handleDragStart(e, { cardHolderIndex, cardIndex })}
                    isOpen={card.isOpen}
                    isDraggable={card.isDraggable}
                    symbol={card.symbol}
                    isHighlighted={card.isHighlighted}
                  />
                );
              })}
            </CardHolder>
          );
        })}
      </div>
    </>
  );
}

export default GamePage;
