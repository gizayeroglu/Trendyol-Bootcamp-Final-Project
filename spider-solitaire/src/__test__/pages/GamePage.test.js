import React from "react";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import { configure, shallow, mount } from "enzyme";

import GamePage from "../../pages/GamePage/GamePage";

configure({ adapter: new Adapter() });

describe("Game Page tests", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<GamePage />);
  });

  it("should render", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
