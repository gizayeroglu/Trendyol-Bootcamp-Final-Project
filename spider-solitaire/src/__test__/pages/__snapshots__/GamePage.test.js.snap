// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Page tests should render 1`] = `ShallowWrapper {}`;
